src/jdataview.js

```diff
@@ -1,833 +1,474 @@
-		);
-
-		return littleEndian || length <= 1 ? result : arrayFrom(result).reverse();
-	},
-
-	// wrapper for external calls (do not return inner buffer directly to prevent it's modifying)
-	getBytes: function (length, byteOffset, littleEndian, toArray) {
-		var result = this._getBytes(length, byteOffset, defined(littleEndian, true));
-		return toArray ? arrayFrom(result) : result;
-	},
+		const result = new Uint8Array(this.buffer, byteOffset, length);
 
-	_setBytes: function (byteOffset, bytes, littleEndian) {
-		var length = bytes.length;
+		return littleEndian || length <= 1
+			? result
+			: arrayFrom(result).reverse();
+	}
 
-		// needed for Opera
-		if (length === 0) {
-			return;
-		}
+	#setBytes(byteOffset, bytes, littleEndian) {
+		const length = bytes.length;
 
-		littleEndian = defined(littleEndian, this._littleEndian);
-		byteOffset = defined(byteOffset, this._offset);
+		littleEndian ??= this.littleEndian;
+		byteOffset ??= this.#bytePointer;
 
-		this._checkBounds(byteOffset, length);
+		this.#checkBounds(byteOffset, length);
 
 		if (!littleEndian && length > 1) {
 			bytes = arrayFrom(bytes, true).reverse();
 		}
 
 		byteOffset += this.byteOffset;
 
-		if (this._isArrayBuffer) {
-			new Uint8Array(this.buffer, byteOffset, length).set(bytes);
-		}
-		else {
-			if (NODE && this._isNodeBuffer) {
-				new Buffer(bytes).copy(this.buffer, byteOffset);
-			} else {
-				for (var i = 0; i < length; i++) {
-					this.buffer[byteOffset + i] = bytes[i];
-				}
-			}
-		}
-
-		this._offset = byteOffset - this.byteOffset + length;
-	},
+		new Uint8Array(this.buffer, byteOffset, length).set(bytes);
 
-	setBytes: function (byteOffset, bytes, littleEndian) {
-		this._setBytes(byteOffset, bytes, defined(littleEndian, true));
-	},
+		this.#bytePointer = byteOffset - this.byteOffset + length;
+	}
 
-	getString: function (byteLength, byteOffset, encoding) {
-		if (NODE && this._isNodeBuffer) {
-			byteOffset = defined(byteOffset, this._offset);
-			byteLength = defined(byteLength, this.byteLength - byteOffset);
+	/**
+	 * Get raw bytes. If length is undefined, it will go to the end of the buffer.
+	 */
+	getBytes(length, byteOffset, littleEndian, toArray) {
+		const result = this.#getBytes(length, byteOffset, littleEndian ?? true);
+		return toArray ? arrayFrom(result) : result;
+	}
 
-			this._checkBounds(byteOffset, byteLength);
+	/**
+	 * Directly set raw bytes at `byteOffset` or the current pointer.
+	 */
+	setBytes(byteOffset, bytes, littleEndian) {
+		this.#setBytes(byteOffset, bytes, littleEndian ?? true);
+	}
 
-			this._offset = byteOffset + byteLength;
-			return this.buffer.toString(encoding || 'binary', this.byteOffset + byteOffset, this.byteOffset + this._offset);
-		}
-		var bytes = this._getBytes(byteLength, byteOffset, true);
+	/**
+	 * Read a string using the specified encoding, or binary if unspecified
+	 */
+	getString(byteLength, byteOffset, encoding) {
```

Comment:
Let's simplify string ops to assume TextEncoder/TextDecoder are also always available.

Reply:
Do we want to keep binary as a separate option? I'm not super familiar with TextEncoder/TextDecoder, but from glancing at the MDN page, it doesn't seem to support it.

Reply:
Yeah binary would still need to be supported manually, I'm only talking about custom fallback for UTF-8.